### DISCUSSION TITLE
`version/info Error` with OIDC

### DISCUSSION THREAD

**Discussion by smolinari** [argoproj/argo-workflows#12070] [Q&A]
Hi,

Does anyone else get a version info error when loading Argo Workflows and haven't yet logged in via SSO/ OIDC?

![image](https://github.com/argoproj/argo-workflows/assets/3314957/8909f867-2696-4dad-abab-8e0ea9cc7274)

Once I've logged in and log back out, it no longer happens. And since logging in works properly, I don't believe it is a setup issue, as noted in the link to the faq. 

I was wondering if this should be reported as a bug?

My take is, if version info is thought of as confidential, then just don't show the version info until the user logs in. Showing the error would then be superfluous. 

Scott

**Comment by agilgur5** [comment]
A duplicate bug report was filed for this in #12168.

> My take is, if version info is thought of as confidential, then just don't show the version info until the user logs in. Showing the error would then be superfluous.

It's not just the version (the message says "version/info"), it's also trying to retrieve user info (`/userinfo` endpont), UI customizations (`/info` endpoint), etc. Depending on your configuration, those may or may not be available to logged out users.

**Comment by atropos112** [comment]
Since this is the earlier topic, I will continue here. 

@agilgur5 your claim seems to be that its failing to call the endpoints and therefore its showing the error ? But not having access to those endpoints before logging in is normal behaviour no? Showing errors when nothing is actually wrong is almost always a bad idea, first you make an unaware user concerned that something is broken even though it isn't, second you teach the user "these pop ups, just ignore them, they pop up for any reason" so when actual problem starts coming up the habit at that point is to ignore them. 

Another question to ask here, who's error is it? Is it a user error , or a server error. If its a user error could you please tell me what error I have made and how do I avoid it in the future? If its a server error then its a bug.

This should be a warning in the console at best.

**Reply by agilgur5** [reply]
To be clear, I didn't write the code and it is [~3 years old](https://github.com/argoproj/argo-workflows/blame/80cd2ec6cffa5e33f8cce67435ba96e68a2a32b4/ui/src/app/app-router.tsx#L54) at this point (which is well before I even used Argo Workflows).

> But not having access to those endpoints before logging in is normal behaviour no?

This part actually gets more hairy since one of the requests is for user info, as I mentioned in my earlier comment. You definitely should have access to the `/userinfo` endpoint before logging in as it needs to be determined in some way whether your token is valid or not.

"If the token is invalid, should we show a message?" is actually a more nuanced question.

> Showing errors when nothing is actually wrong is almost always a bad idea, first you make an unaware user concerned that something is broken even though it isn't, second you teach the user "these pop ups, just ignore them, they pop up for any reason" so when actual problem starts coming up the habit at that point is to ignore them.

For sure, "alert fatigue" etc is a problem. 100% agree. I mentioned something similar in https://github.com/argoproj/argo-workflows/issues/10826#issuecomment-1704715649 as well.

> Another question to ask here, who's error is it? Is it a user error , or a server error. If its a user error could you please tell me what error I have made and how do I avoid it in the future? If its a server error then its a bug.

I describe this in more detail in my recent comment on your issue: https://github.com/argoproj/argo-workflows/issues/12168#issuecomment-1802928934. It is a notification, which does not necessarily mean that you did anything wrong.

But if you have a misconfiguration, that would indeed mean that Argo was misconfigured, and that's why you're getting a "token invalid" error with a link to the FAQ on how to debug that.

> This should be a warning in the console at best.

The misconfiguration notice/FAQ link I suppose could be a warning in the console, although that's not necessarily ideal either. It's a common misconfiguration, and an operator of Argo may not even know to look at the console for such a warning 😕 
I happen to both be very experienced in front-end and SRE and so know my way around both the browser console and k8s well, but this certainly does not apply to all operators. 
The misconfiguration notice/FAQ link is for a different audience than all end-users; it is primarily for operators debugging a recent SSO configuration.

I agree that this notification should be of type warning not of type error, as I wrote in my comment on your issue: https://github.com/argoproj/argo-workflows/issues/12168#issuecomment-1802928934. Whether it should just be in the console is a bit trickier given that it could just be missed entirely there. An ideal may perhaps be somewhere in between

> Since this is the earlier topic, I will continue here.

Welp, I didn't see this until after I responded to your issue. I tried to answer some of the pieces that are different here than there. Issues are the proper place for enhancement requests and bug reports, but discussions make more sense for say, understanding how the logic works.

**Reply by smolinari** [reply]
The requirement that a user be logged in to see the version and it end up an error for any user that isn't logged in is a bug in my book. Glad you wrote it up @atropos112 :+1: 

@agilgur5 

You wrote in the issue:

> so to improve this we would have to attempt to distinguish between a misconfiguration and a plain invalid token, which is correct behavior.

Everything in my installation of Argo Worflows is working as it should. SSO works perfectly with Keycloak OIDC. Yet, I still am getting the error message. That means, it is showing due to an invalid token, correct? If that is the case, then that should be a console error and I should be seeing the login page only. If I want to see why I can't login in, I'll dig deeper i.e. look into the console. A "normal" user of such a system would need to report back to me they can't login in, so they don't need the warning either. 

I have a similar discussion with another team about their system showing an OAuth token being invalid to the users, when the OIDC refresh token times out. It's a lot more descriptive, but it is also totally unnecessary. Just show the login page, because the user is technically logged out. Why that token is invalid can be for many reasons, misconfiguration only being one of them. And, as a sys-admin, if users can't log in, I'll have to dig deeper as to why i.e. look into why tokens aren't validating like looking for console errors or system logs. The warning/ error toast is completely superfluous.

Also, at the risk of repeating myself, as I read the warning/ error, it seems more like it is arising because the requirement is to deem version info as confidential to only logged in users. However, if I log in and log back out, the error doesn't happen. So, that just adds confusion to the problem. At any rate, if the version info is confidential for logged out users, then simply don't show the version info on the login page. 

This is definitely a bug. And more like a blemish to a polished software than anything serious. But, even a blemish on a diamond can make the diamond at lot less attractive and less valuable. I hope you can fix this and thanks for your attention. :grin: 

Scott

**Reply by atropos112** [reply]
First of all thank you @agilgur5 for considering this so carefully and and @smolinari for looking into this.

I agree with @smolinari  that you will have only two types of users:
1. Typical users, the ones who use the application and have nothing to do with setting it up or tweaking it 
2. Sysadmins, individuals who will dig deep into logs and console to tweak it and make it work

Lets consider the possible situations:

- If there is an issue with token due to missconfiguration typical users will bother sysadmins, but realistically not being able to login using SSO would "ring the bells" regardless if you see such error or not, so having this error brings no actual value here.
- If there is no missconfiguration issue they will still bother sysadmins with "why do i get this error", causing fatigue in users, and wasting sysadmins time (thats how i got here).

If there is infact an actual missconfiguration issue, as outlined above sysadmin's will be aware that there is an issue (as users can't login via SSO and users will be much louder than this error can ever be) and they will investigate. If a sysadmin investigates i think it takes sub 1 minute for such individual to think "let me check the console", this error will save seconds at best.

Alternatively revisiting the idea from #10826 i think it might be worth having an environment variable for reducing warnings/errors to minimum, as a sysadmin I would much rather have this on.

**Reply by agilgur5** [reply]
> That means, it is showing due to an invalid token, correct?

~Correct, probably an expired token if I had to guess~ **EDIT**: see [below comment](https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-10108986), you'd get a different `message` from "token not valid" in that case, though it would still be after the "Failed to load version/info Error: ".

> Just show the login page, because the user is technically logged out.

I know some websites do this, and I did get quite accustomed to this working in an enterprise environment where no token lasts for more than 8 hours (if that), but IMO it is a bit surprising as a user when you're kicked back to the login screen with no explanation what-so-ever.
Perhaps we as developers are more familiar with the context though and so are less surprised. I am unsure that that applies to all audiences of Argo equally. For example, I worked on a ML Platform, and some users were pure analysts with little development experience. In that example, I'm not sure an analyst would be familiar with how tokens work/expire to know why they got kicked back to the login screen.

> Why that token is invalid can be for many reasons, misconfiguration only being one of them.

Yea absolutely. It is unfortunately a pretty _common_ misconfiguration though 😅 Unfortunately Argo can't really distinguish which scenario it is for you 😕 

> And, as a sys-admin, if users can't log in, I'll have to dig deeper as to why i.e. look into why tokens aren't validating like looking for console errors or system logs.

For sure. The Server logs were improved in #11370 and the FAQ link actually appears in the back-end error. It just so happens that this front-end code [directly prints out](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/ui/src/app/app-router.tsx#L54) the back-end error in the toast.

I had a good bit of trouble getting RBAC working (as #10927, my first ever PR to Workflows, mentions) and a smaller bit on SSO (mostly the Dex integration portion). The lack of visible UI errors around RBAC was pretty confusing as per #8760. It's just "not working" and you have to do a lot of inference as to why (eventually, as in #10927, I figured out a way to get the Server to print some useful information). Ideally, we make errors as clear as possible so there isn't a need to guess-and-check, but yes, there is a limit to how much we can do there.
Striking the proper balance for both the initial set-up debugging experience and the general user experience has some challenges. Kind of similar to how you structure your logs and decide what's an `info` level, what's a `debug` level, etc. There's some subjectivity there depending on who you talk to.

Personally, I don't believe I checked the console much around auth errors, as those are on the back-end, so I'm not even sure if I would  have read a console warning. And for context, I am pretty experienced in front-end, given that I built one of the first production websites to use React back in early 2014 and maintain several front-end libraries and TS compiler integrations. 

> it seems more like it is arising because the requirement is to deem version info as confidential to only logged in users.

it's also just about the token being invalid -- the version info is picked up in [the same request chain](https://github.com/argoproj/argo-workflows/blob/80cd2ec6cffa5e33f8cce67435ba96e68a2a32b4/ui/src/app/app-router.tsx#L63). It's just one error for the whole chain. So even if we didn't make a request for the version info, it would still show up from the `/userinfo` request.

As in, it's more than just not making the version info request. I would like to solve all of the above at once, not just that portion.

> However, if I log in and log back out, the error doesn't happen.

Yea that one I was a bit confused by 🤔 
I thought perhaps if it's an internal redirect in the SPA (Singe Page App), then it didn't re-render the component when manually logged out and so didn't make the request again. But the [`logout` function](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/ui/src/app/login/components/login.tsx#L10) appears to refresh the page entirely, so the same requests _should_ be made. 
~The other possibility is that may be because you have _no token at all_ anymore, so the authentication [skips](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/server/auth/gatekeeper.go#L192) the token validation check. Or rather, no token is "valid", while an expired token is "not valid".~ **EDIT**: the authentication doesn't skip no token actually, that's why the error happens in the first place before you're logged in. Not sure how this happened. Also see [below comment](https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-10108986), an expired token is a slightly different error that doesn't say "token not valid".

> 2\. Sysadmins

For clarity, the term in the k8s ecosystem is "Operators" or cluster-admins, which is why I used that term. That's also why docs often have an "Operator Manual" separate from the "User Guide", as [Argo does](https://argoproj.github.io/argo-workflows/installation/). Automation in the space, i.e. code as in "the Operator framework", are also sometimes called "Operators" intentionally because they automate some of a cluster-admin's work (basically automated vs. manual operations).

> Alternatively revisiting the idea from #10826 i think it might be worth having an environment variable for reducing warnings/errors to minimum, as a sysadmin I would much rather have this on.

#10826 mentions this too -- we might as well remove these if we disable them by default. Too many optional pieces can become a maintenance headache too. I'd rather get the default right for everybody.

<hr>

Going back to my proposal from https://github.com/argoproj/argo-workflows/issues/12168#issuecomment-1802928934, how would both of you feel if the toast was changed to a "Warning" type (it really shouldn't be an "Error") and it said "token not valid. See logs for more details if you believe this to be a misconfiguration"? That feels awfully verbose and still confusing though, I don't particularly like it.
Another option is just saying "token not valid" in a warning without elaboration and _assuming_ that someone will check the logs or console if they think this is a misconfiguration. But it sounds like both of you are saying that "token not valid" in general is confusing?

We could potentially assume "token expired, please login again", but that's not correct in all cases, though is likely correct in most cases. Would like to have something in between those options ideally 😕 **EDIT**: actually expiration specifically is checkable, see below, but some SSO scenarios are hard to catch

**Reply by smolinari** [reply]
>  I am unsure that that applies to all audiences of Argo equally.

If there is an unexpected "logout" by a user who isn't "in the know", they'll hit up their sys-admin, ehem..... operator, with their surprise. The error won't help that user. The sys-admin will. Leave it up to the sys-admins. I've not seen one system other than Atgo Workflows and that one other application I mentioned, that does this kind of error for invalid OIDC tokens. 

> Unfortunately Argo can't really distinguish which scenario it is for you

And it shouldn't. It should offer, at best, a logging of the invalid token. For the user, just show her logged out/ the login screen. That is all that is necessary. 

> But it sounds like both of you are saying that "token not valid" in general is confusing?

Correct! Just log the user out/ keep the user logged out. But, do log the access denial somewhere and as descriptive as can be. Just not in a toast in the UI. We'll figure out the issue from there, IF it even is an issue (maybe the token is just timed out).

> We could potentially assume "token expired, please login again", but that's not correct in all cases, 

Correct. Thus why you keep the user logged out with no other warnings in the UI, because you can't decipher if it is a real issue or just a timed out token. What is 100% certain is the user isn't logged in. And that is how it should be. :grin: 

Scott

**Reply by agilgur5** [reply]
### getting back to this

> A duplicate bug report was filed for this in #12168.

So apparently both this discussion and #12168 are both duplicates of an older issue, #7921.

It took me a while to analyze this situation as I thought the logic must be a bit off that we can't detect an expired token or other scenarios. I didn't want to make a half-baked, potentially further problematic (or problematic in different ways) solution to this, so I needed some time to deep-dive. This specific error has been iterated on before my time as a contributor too and continues to have issues about it, so it deserves a very thorough look-over, root cause analysis, fixes etc. Not to mention it impacts nearly every user as it's login (whereas most bugs tend to be for specific use-cases).
I'm also a volunteer, simultaneously have been the top contributor for ~a year, and we [lack contributors](https://github.com/argoproj/argo-workflows/blob/28dc65e/community/sustainability_effort.md), so uh, suffice to say, I have about two helpings more than enough on my plates with all the work for this project 🫠 (and I maintain other things too etc etc)

It turns out my assumption was partially right. While we can't detect what auth modes you have enabled yet in the UI (see #8179), we can detect some errors, although the logic is pretty nuanced, particularly due to multiple auth modes and multiple headers (#7607).

### corrections

I actually need to correct some of my old statements now that I have a more thorough understanding as they were not correct:
> ~Correct, probably an expired token if I had to guess~

An expired token actually has a [slightly different error](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/server/auth/gatekeeper.go#L209). The UI would still show it "Failed to load version/info Error:" but the `message` wouldn't be "token not valid" and would instead have [a prefix](https://github.com/argoproj/argo-workflows/blob/52cca7e079a4f6d76db303ac550b1876e51b3865/server/auth/sso/sso.go#L369) saying "failed to validate claims:".

> ~The other possibility is that may be because you have _no token at all_ anymore, so the authentication [skips](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/server/auth/gatekeeper.go#L192) the token validation check. Or rather, no token is "valid", while an expired token is "not valid".~

I must've read this wrong or something at the time (I've looked at it way too many times), but the API does not decipher "no token at all" and just considers it an invalid token too. That should probably be changed to a more specific error like "no token provided" or something.

For the expired token, see the above paragraph

### moving forward

I'm working on detecting these different error scenarios in the UI, which also necessitate some API changes. The varying auth modes make this surprisingly complicated since they work quite differently; I've had many "oh I can just do this" moments only to realize "oh wait that doesn't work for this scenario for this auth mode".

For instance:
- The API checks your token against [all configured auth modes](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/server/auth/gatekeeper.go#L184) and has some hacks for [`server` mode](https://github.com/argoproj/argo-workflows/blob/873bacbafc4da79df15e0da728501ad925eef4fc/server/auth/gatekeeper.go#L177).
- A token failing for one mode does not instantly return an error, as it will try other modes -- you don't request a specific mode with your token. This is inferable from the token, but the API and UI don't currently check this, so this is another clear area to refactor.
- `server` mode is effectively unauthenticated, and in that case, you don't need a token at all.
- `client` and `sso` have [two different tokens](https://github.com/argoproj/argo-workflows/blob/52cca7e079a4f6d76db303ac550b1876e51b3865/server/auth/mode.go#L33) with different validation and different authorization flows.
  - `client` will have a token when attempting to login (as you input it), but `sso` may not (if the SSO failed).
  - `sso` causes a full page reload due to redirects, but the others do not necessarily do that.
- There's no "login" API method unlike traditional user-based systems, so quite literally this exact "version/info" check is used as a proxy for detecting whether one is authenticated or not. Which is confusing from many perspectives and takes a bit to wrap your head around.

Basically I need to refactor the UI and API to infer the auth modes, give more specific errors, and then act differently based on those errors.
If the former two had been done already, this wouldn't have been as confusing, although the lack of "login" method is confusing in its own right as well. I suspect that some of these errors got more ambiguous as more auth modes, multiple, simultaneous auth modes, and multiple headers were added -- i.e. years of tech debt (like many other parts of the codebase; I've done quite a lot of upgrades/refactoring/clean up/etc as a contributor).

Since the "version/info" check is used as a proxy, it's always possible to get an error. Then we need to figure out whether that error makes sense to display and what it should say:
1. obviously if you haven't even logged in yet, we shouldn't display it
    1. an empty cookie can suggest you haven't
        1. but an SSO failure could result in no cookie either.
            1. I believe [these SSO errors](https://github.com/argoproj/argo-workflows/blob/52cca7e079a4f6d76db303ac550b1876e51b3865/server/auth/sso/sso.go#L258) aren't necessarily caught in the UI during the redirect back. If they were, I think we could skip trying to authenticate entirely, not just skip displaying the error.
        1. and `server` mode would have no cookie, but also shouldn't fail authentication
    1. if you do have a cookie, then you tried logging in, and your cookie is invalid for various reasons
        1. we should display some form of "Login Failed" message, which is a bit more appropriate for non-operator end users of the UI
            1. and then a "sub-error" that is more specific about what is invalid with the cookie
            1. the docs link can probably be left server-side as a warning "in case of misconfiguration" and not show UI-side, as that seems to one of the bigger complaints
            1. based on your cookie's prefix, we can detect which auth mode you tried

**Reply by agilgur5** [reply]
The [above](https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-10108986) is being implemented in #13372. Currently a draft, but ~80% done. Still needs SSO error detection and a bunch of testing.

**Comment by Crayeth** [comment]
I am maintaining an argo workflow setup and this error popup is scary and confusing for my users and raises questions I don't want to spend my time on. Ideally I would like to at least be able to turn it off somehow.

**Reply by omar-khaled-mahmoud** [reply]
Any updates?

**Comment by crdsvc** [comment]
@agilgur5 is there a way to disable the popup before the user logs in ? Any workarounds ?

**Reply by agilgur5** [reply]
No. If there were, it would've been mentioned already.
I'm [working on a fix](https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-10108986) to only show the popup when appropriate.

**Reply by smolinari** [reply]
If I recall the issue correctly, the "version/info", if only accessible after login or rather it can only be seen by someone who has logged in at some point successfully, then it should be removed from the login page for those who can't see it, instead of offering the useless.even waste causing error. So, I don't understand the complexity involved. It's only complex, when one tries to show/ get "version/info" and can't for whatever reasons. Just don't show this data on the login page. Is that too simple?

Scott

**Reply by agilgur5** [reply]
Again, it's not just that, the same request chain is used as an auth check.

As I said before in https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-7516017, https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-7537069, and in my most recent comment https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-10108986:

> it's also just about the token being invalid -- the version info is picked up in [the same request chain](https://github.com/argoproj/argo-workflows/blob/80cd2ec6cffa5e33f8cce67435ba96e68a2a32b4/ui/src/app/app-router.tsx#L63). It's just one error for the whole chain. So even if we didn't make a request for the version info, it would still show up from the `/userinfo` request.

> There's no "login" API method unlike traditional user-based systems, so quite literally this exact "version/info" check is used as a proxy for detecting whether one is authenticated or not. Which is confusing from many perspectives and takes a bit to wrap your head around.

I wish it were that simple. And if it were, not only would I have fixed it already (and anyone could have PR'd a fix without needing substantial context), it would probably have been fixed well before I was even a contributor. This error has existed for _years_ (it predates #7921)

This is also the wrong subthread

**Comment by smolinari** [comment]
Sorry, I entered the reply in the wrong field. 

So, the error is arising, because the user is "not logged in" or hasn't been logged in via OIDC, correct? Or put another way, the "check" of the token(s) is failing, which is actually a correct behavior for someone who isn't logged in? If yes, to the user, she should simply be seeing a login page shown with no error, as the user hasn't attempted to log in. For the sys-admin, the error should be visible in the console. This allows troubleshooting when login attempts are failing.  

Going a step back, if there is a proper login error happening, i.e. the user it trying to login and just keeps getting the login page because the "handover" via OIDC fails (and not because of false credentials or the like), then this "I keep getting the login page" will be reported to the sys-admin, who then can see the error in the console. So, an initial token check resolves to a route to the login page, with the result of the token check (the error) in the console. Version/Info data is not shown on the login page and the user might be perplexed, but has a proper issue to report. If it was just an initial route to the login page and logins are working properly, then no error should be visible.  

I'm trying to describe the user's experience in this. I had a similar application where they too put up an error to the user that a token wasn't valid, but before the user attempted to log in (again). On my complaint, they changed the process to just log the user out and only show the login page (with the button for OIDC login to log in). This is the same situation it seems and I believe needs the same resolution.

Scott

**Reply by agilgur5** [reply]
> So, the error is arising, because the user is "not logged in" or hasn't been logged in via OIDC, correct? Or put another way, the "check" of the token(s) is failing, which is actually a correct behavior for someone who isn't logged in? 

Yes. It's not specific to SSO/OIDC though.

> If yes, to the user, she should simply be seeing a login page shown with no error, as the user hasn't attempted to log in. For the sys-admin, the error should be visible in the console.

You suggested this [before](https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-7537276) and that is part of the solution I added in my PR #13372

> then this "I keep getting the login page" will be reported to the sys-admin, who then can see the error in the console.

There's two issues with this unfortunately:
1. It doesn't just affect SSO. If you login via `client` auth, i.e. an SA token, and your token is invalid, you'll get the same error. In this case, you're hitting "login" and going back to the login page each time. If there's no error, that's super confusing. And in this case, the logs will similarly just generically say "token not valid" so a sysadmin/operator may not be able to help much either (there are more logs in the SSO case, i.e. #11370)
2. "I keep getting the login page with no error" is very confusing too, and based on my experience, I can pretty much guarantee that would get plenty of issues and support questions in Slack too. Even with sysadmins/operators, I've had to tell folks to check the rest of their logs with some frequency (i.e. #11370), which tends to reveal the problem 😕

Can't solve folks not looking at their logs ofc, but we can help the folks that do to understand the problem/misconfiguration. #13372 makes the logs more detailed too for an invalid token.

Mitigating these problems _wholesale_ will reduce the support burden and help with self-resolution. The complexity partly results from multiple auth modes, as I wrote [previously](https://github.com/argoproj/argo-workflows/discussions/12070#discussioncomment-10108986) (under the "moving forward" section), where making one case better can be at the cost of another case if not done carefully (I quite literally did this several times in early development, per the second sentence in that section).
That and the whole lack of a "login" API, as I also noted in that section, which I imagine is due to legacy since the UI used to be a standalone tool used locally, not hosted in the same cluster, and even in its own repo. It can still be used locally and does not need hosting necessarily, where the auth is generally equivalent to `server` auth mode and no login is necessary.
Also SSO is a bit tedious to test locally so that adds a non-trivial amount of time too. Plus the iteration costs of restarting the Server, clearing cookies, etc.
The actual PR is pretty small as a result, but getting it to handle all cases and testing and iterating was/is complicated.

**Reply by smolinari** [reply]
Well, no matter. I'm glad you are on it and have a handle on it too. I'm just a power user and sys-admin and not really a (great) developer. Thanks for getting on this and looking forward to the fixes. 

Scott

**Reply by agilgur5** [reply]
Thanks for your earlier suggestions, it did help me realize that there must be something missing in the code that we can't just remove the message, or even further, before login, not make a request at all. And indeed there were things missing that made that not possible mostly because of multiple auth modes and no "login" API and the like. Once I inferred them (both in my head and in the code), that made things _a lot_ easier, but there's still a lot of conditionals that can take a bit to wrap your head around too.
I also found at least two other bugs in the process that I'm fixing too 😅
